File: site/lib/tasks.ts
"use server";

import { auth } from "@/auth"
import { Task } from "@prisma/client"

const API_URL = process.env.API_URL

export async function GetTasks(projectId: string) {
    const session = await auth()

    const url = new URL(`/projects/${projectId}/tasks`, API_URL)
    if (projectId) {
        url.searchParams.set("projectId", projectId)
    }

    return fetch(url, {
        headers: {
            "Content-Type": "application/json",
            "Authorization": session.sessionToken
        }
    }).then(res => res.json() as Promise<Task[]>)
}

export async function CreateTask(projectId: string, data: { title: string, description: string, dueDate?: string | null, status: string }) {
    const session = await auth()

    const url = new URL(`/projects/${projectId}/tasks`, API_URL)

    return (await fetch(url, {
        method: "POST",
        headers: {
            "Content-Type": "application/json",
            "Authorization": session.sessionToken
        },
        body: JSON.stringify(data)
    })).ok
}

export async function UpdateTask(projectId: string, taskId: string, update: Partial<Task>): Promise<{
    ok: boolean,
    data: Task
}> {
    const session = await auth()

    const url = new URL(`/projects/${projectId}/tasks/${taskId}`, API_URL)

    console.log(update)

    const result = await fetch(url, {
        method: "PATCH",
        headers: {
            "Content-Type": "application/json",
            "Authorization": session.sessionToken
        },
        body: JSON.stringify(update)
    })

    return {
        ok: result.ok,
        data: await result.json()
    }
}

export async function DeleteTask(projectId: string, taskId: string) {
    const session = await auth()

    const url = new URL(`/projects/${projectId}/tasks/${taskId}`, API_URL)

    return (await fetch(url, {
        method: "DELETE",
        headers: {
            "Authorization": session.sessionToken
        }
    })).ok
}

File: tasks.py
import datetime
from typing import Optional

from prisma import Prisma
from pydantic import BaseModel

class ExecuteTaskBody(BaseModel):
    title: str
    description: str
    dueDate: Optional[datetime.datetime]
    status: str

class ExecuteCompleteTaskBody(BaseModel):
    title: str
    description: str
    dueDate: Optional[datetime.datetime]
    status: str
    completed: bool

class TasksHelper:
    def __init__(self, prisma: Prisma):
        self.prisma = prisma

    async def create_task(self, project_id: str, user_id: str, title: str, description: str, dueDate: Optional[datetime.datetime], status: str):
        return await self.prisma.task.create(data={
            "title": title,
            "description": description,
            "dueDate": dueDate,
            "status": status,
            "projectId": project_id,
            "userId": user_id
        })

    async def update_task(self, project_id: str, user_id: str, task_id: str, title: str, description: str, dueDate: Optional[datetime.datetime], status: str, completed: bool):
        return await self.prisma.task.update(
            where={
                "id": task_id,
                "projectId": project_id,
                "userId": user_id
            },
            data={
                "title": title,
                "description": description,
                "dueDate": dueDate,
                "status": status,
                "completed": completed
            }
        )

    async def delete_task(self, project_id: str, user_id: str, task_id: str):
        return await self.prisma.task.delete(where={
            "id": task_id,
            "projectId": project_id,
            "userId": user_id
        })

    async def get_tasks(self, project_id: str):
        return await self.prisma.task.find_many(
            where={
                "projectId": project_id
            },
            order={
                "createdAt": "desc"
            }
        )